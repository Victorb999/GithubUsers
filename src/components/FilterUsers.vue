<template>
  <div class="filterContainer">
    <form @submit.prevent="setFilter" class="form">
      <input
        type="text"
        name="search"
        placeholder="Escreva aqui o nome de um usuÃ¡rio"
        v-model="state.name"
      />
      <button type="submit" class="btn-primary">Buscar</button>

      <div class="filter">
        <span v-if="state.nameValidation != ''">{{
          state.nameValidation
        }}</span>
      </div>
      <div class="filter">
        <input
          type="checkbox"
          v-model="state.respositories"
          name="repositories"
        />
        <label for="repositories">
          Buscar somente quem tem mais de 20 repositÃ³rios</label
        >
      </div>
      <div class="filter">
        <input type="checkbox" v-model="state.followers" name="followers" />
        <label for="followers">
          Buscar somente quem tem mais de 20 seguidores</label
        >
      </div>
      <div class="resetFilter">
        <button
          type="button"
          title="Busca os 1000 primeiros brasileiros com mais followers"
          class="reset"
          @click="resetFilter"
        >
          ðŸ”ƒ resetar busca
        </button>
      </div>
    </form>
  </div>
</template>

<script lang="ts">
import { defineComponent, reactive } from "vue";
import { useStore } from "vuex";
export default defineComponent({
  name: "FilterUsers",
  setup() {
    const store = useStore();

    interface Data {
      name: string;
      nameValidation: string;
      respositories: boolean;
      followers: boolean;
    }

    const state = reactive({
      name: "",
      nameValidation: "",
      respositories: false,
      followers: false
    }) as Data;

    function setFilter() {
      const MIN_SIZE_NAME = 3;
      const MIN_SIZE_FILTER = 20;
      if (state.name.length > MIN_SIZE_NAME) {
        state.nameValidation = "";
        let filter = [state.name + " in:login"];
        if (state.respositories === true) {
          filter = [...filter, `repos:>${MIN_SIZE_FILTER}`];
        }
        if (state.followers === true) {
          filter = [...filter, `followers:>=${MIN_SIZE_FILTER}`];
        }
        store.dispatch("changeQuery", filter);
      } else {
        state.nameValidation =
          "Tente digitar um nome de usuÃ¡rio com mais que 3 letras.";
      }
    }

    function resetFilter() {
      const filter = ["location:BRASIL", "location:BRAZIL"];
      state.name = "";
      state.nameValidation = "";
      state.respositories = false;
      state.followers = false;
      store.dispatch("changeQuery", filter);
    }

    return { state, setFilter, resetFilter };
  }
});
</script>

<style lang="scss" scoped>
.filterContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  input[type="text"] {
    border-radius: 5px;
    min-width: 30vw;
    border: none;
    border-radius: 5px;
    padding: var(--size-2) var(--size-3);
    margin: var(--size-2) var(--size-1);
    box-shadow: 0 0.5px var(--size-1) var(--gray-3);
    font-weight: 400;
    color: var(--dark);
    &:focus {
      outline: none;
    }
  }

  button {
    padding: var(--size-2) var(--size-3);
    margin: var(--size-2) var(--size-1);
  }
  span {
    display: flex;
    flex-wrap: wrap; // Quebra a linha
    flex-direction: row; //DireÃ§Ã£o que vocÃª deseja
    max-width: 90vw;
    margin-top: var(--size-3);
    color: var(--primary);
    font-weight: bold;
  }
  .filter {
    display: flex;
    align-self: center;
    align-items: center;
    justify-content: center;
    max-width: 90vw;
    margin-top: var(--size-3);
    input {
      margin-right: var(--size-2);
    }
  }
  .reset {
    padding: var(--size-2) var(--size-3);
    margin: var(--size-2) var(--size-1);
    background: transparent;
    &:hover {
      background: var(--gray-2);
    }
  }
  .resetFilter {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-top: var(--size-3);
  }
}
.theme-dark {
  .filterContainer {
    input[type="text"] {
      box-shadow: 0.5px 0.5px var(--size-1) var(--aux);
      background-color: darken(#1f1b24, 10%);
      color: var(--white);
    }
    .reset {
      color: var(--white);
      transition: all 0.5s;
      &:hover {
        box-shadow: 0.5px 0.5px var(--size-1) var(--aux);
        background-color: darken(#1f1b24, 10%);
      }
    }
  }
}
@media (max-width: 900px) {
  .filterContainer {
    input[type="text"] {
      min-width: 50vw;
    }
  }
}
@media (max-width: 400px) {
  .filterContainer {
    form {
      display: flex;
      flex-flow: row wrap;
      align-items: center;
      justify-content: center;
    }
    input[type="text"] {
      min-width: 90vw;
    }
  }
}
</style>
