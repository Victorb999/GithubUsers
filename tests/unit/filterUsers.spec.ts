import { shallowMount } from "@vue/test-utils";
import FilterUsers from "@/components/FilterUsers.vue";
import { createStore } from "vuex";

describe("Filter Users", () => {
  const store = createStore({
    state() {
      return {
        query: ["location:BRASIL", "location:BRAZIL"]
      };
    },
    mutations: {
      changeQuery(state: any, payload: any) {
        state.query = payload;
      }
    }
  });

  const build = () => {
    const wrapper = shallowMount(FilterUsers, {
      global: {
        plugins: [store]
      }
    });
    return wrapper;
  };
  it("renders the component", () => {
    // arrange
    const wrapper = build();
    // assert
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders filter", () => {
    // arrange
    const wrapper = build();
    const filter = wrapper.find(".filterContainer");
    expect(filter.exists()).toBe(true);
  });

  // it("filter less 3 caracters", async () => {
  //   const wrapper = build();
  //   const textInput = wrapper.find('input[type="text"]');
  //   await textInput.setValue("12");
  //   await wrapper.find("button").trigger("click");

  // });
});
