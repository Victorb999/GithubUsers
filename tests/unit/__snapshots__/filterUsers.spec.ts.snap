// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Users renders the component 1`] = `
<div class="filterContainer">
  <form><input type="text" name="search" placeholder="Escreva aqui o nome de um usuÃ¡rio"><button type="submit" class="btn-primary">Buscar</button>
    <!--v-if-->
  </form>
</div>
`;
